src/notifications/common.ts

```diff
@@ -35,3 +45,66 @@ export enum NotificationPreferenceType {
 export enum NotificationPreferenceStatus {
   Muted = 'muted',
 }
+
+export const notificationPreferenceMap: Partial<
+  Record<NotificationType, NotificationPreferenceType>
+> = {
+  [NotificationType.ArticleNewComment]: NotificationPreferenceType.Post,
+  [NotificationType.CommentReply]: NotificationPreferenceType.Comment,
+  [NotificationType.SquadPostAdded]: NotificationPreferenceType.Source,
+  [NotificationType.SquadMemberJoined]: NotificationPreferenceType.Source,
+};
+
+type NotificationPreferenceUnion = NotificationPreferenceComment &
+  NotificationPreferencePost &
+  NotificationPreferenceSource;
+
+export const saveNotificationPreference = async (
+  con: DataSource | EntityManager,
+  userId: string,
+  referenceId: string,
+  notificationType: NotificationType,
+  status: NotificationPreferenceStatus,
+) => {
+  const type = notificationPreferenceMap[notificationType];
+
+  if (!type) {
+    throw new ValidationError('Notification type not supported');
+  }
+
+  const params: Partial<NotificationPreferenceUnion> = {
+    type,
+    userId,
+    status,
+    notificationType,
+    referenceId,
+  };
+
+  switch (type) {
+    case NotificationPreferenceType.Comment:
+      params.commentId = referenceId;
+      break;
+    case NotificationPreferenceType.Post:
+      params.postId = referenceId;
+      break;
+    case NotificationPreferenceType.Source:
+      params.sourceId = referenceId;
+      break;
+  }
+
+  try {
+    await con
+      .getRepository(NotificationPreference)
+      .createQueryBuilder()
+      .insert()
+      .values(params)
+      .orIgnore()
+      .execute();
+  } catch (err) {
+    if (err.code === TypeOrmError.FOREIGN_KEY) {
+      throw new NotFoundError('Invalid reference id');
+    }
+
+    throw new Error('Something went wrong');
```

Comment:
Please throw the original error so we keep info of what happened
```suggestion
    throw err;
```